### PR TITLE
Require `throw` statement in tests for early errors

### PR DESCRIPTION
**Pull request by jugglinmike**
In gh-778, we began tracking tests for early errors with metadata. Prior to
that patch, some tests for early errors began with a `throw` statement. The
intent of this statement was to ensure that program code was not executed.

With the introduction of the metadata, we elected to remove that pattern in
favor of a recommendation for test consumers to dynamically insert such a
statement immediately prior to test execution [1].

However, this approach increases the burden for test consumers, and because it
makes the test material less literal, it also tends to obscure the expected
behavior (see the discussion at gh-878 for more information).

With the introduction of a static analysis tool [2], the project is now in a
position to automatically enforce a requirement for an explicit `throw`
statement. This change set introduces such a requirement and updates all tests
to satisfy it. In so doing, it makes expectations more transparent both for
contributors and consumers, at the expense of a small redundancy in test
content (i.e. tests for early errors must now include dedicated metadata and a
`throw` statement).

The linting heuristic is admittedly a little weak--it only verifies that the
source text contains text for a `throw` statement, not that it is the first
statement in the file (or even that it is a statement at all). Anything more
than this would significantly increase the linter's complexity, but I don't
think the relative benefit warrants this. For now, at least, it seems fair to
expect maintainers to recognize those extenuating circumstances that would fool
the heuristic.

[1] Commit ade6d2e384350055e1e23c35cfbf0d4dab03c82e
[2] Commit 74954bfa915136c7998f8a0d9ca40b885075d409

### CODE REVIEW

**Comment by littledan** [comment]
This seems like a reasonable approach to me; great to see your attention to detail here.

**Comment by rwaldron** [comment]
I wanted to just take @littledan's word for it, but I had to spend some time catching up on the cited patches to make sure the I fully follow this change.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -187,7 +187,7 @@ function $DONE(arg) {
 
 ## Handling Errors and Negative Test Cases
 
-Expectations for **parsing errors** should be declared using [the `negative` frontmatter flag](#negative):
+Expectations for **parsing errors** should be declared using [the `negative` frontmatter flag](#negative). They must also include the exact `throw` statement specified in this example (in order to guarantee that implementations do not execute the code):
```
Comment (rwaldron) [inline CONTRIBUTING.md]: I wonder how we can call more attention to this—maybe some bold text? What does everyone else think? Maybe it's fine as it is?

**Comment by leobalter** [comment]
I would say this should be responsibility of the test runner, prepending early error tests with a throw statement.

This change is already affecting 2000+ files and I wonder if this will raise the cost of reviews: "please add a throw statement before the failing code".

On the other side: there might a lot of test runners we don't have control and can't secure they will implement this approach. I would say this diplomatic communication would be easier.

This documentation can suggest test runners should prepend a throw statement on early error tests. 

What do you think, @jugglinmike, @rwaldron, @littledan?

cc'ing @anba and @bterlson wrt SpiderMonkey and Chakra as well.

**Comment by rwaldron** [comment]
>  this will raise the cost of reviews: "please add a throw statement before the failing code".

The additional cost is minimal, because the work is done in CI. 

> there might a lot of test runners we don't have control and can't secure they will implement this approach.

This is the problem that @jugglinmike is solving directly.

File: tools/lint/lib/checks/negative.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            return '"negative" must be a dictionary with fields "type" and "phase"'
+
+        if not 'type' in negative:
+            return '"negative" must specify a "type" field'
+
+        if not 'phase' in negative:
+            return '"negative" must specify a "phase" field'
+
+        if negative["phase"] == "early" and not _THROW_STMT.search(source):
+            return 'Negative tests of type "early" must include a `throw` statement'
```
Comment (leobalter) [inline tools/lint/lib/checks/negative.py]: oh, I finally got this! Thank you, @jugglinmike! 👑 👑 👑

File: CONTRIBUTING.md (hunk 1 above)
Comment (leobalter) [inline CONTRIBUTING.md]: I think it's fine. What I think we should improve eventually a way for contributors to use the linter before the CI running over the PR. Something that puts it more visible or clear.

**Comment by rwaldron** [comment]
👍

**Comment by bterlson** [comment]
FWIW I also agree that this should be a harness requirement. There are a number of requirements that test runners have in order to fully test ECMAScript. This change feels to me like checking in separate files for strict, strict module, and non-strict mode.

**Comment by leobalter** [comment]
@bterlson this PoV is important as I don't consume Test262 (as an implementor, etc). @jugglinmike uses it on JSHint, so does the other ppl I pinged here. 

I believe if we work on this requirement of runners prepending a throw statement, we would be able to get rid of this manual insertion. For now, I like this is at least checked through the linter.

**Pull request merged**